### PR TITLE
Dielectric charge

### PR DESCRIPTION
**Pull request by keniley1**
This PR adds a surface charge material to Zapdos. The type of ion and electron BC you use needs to be specified, and it defaults to Hagelaar since that is typically what is used. The material uses AD to allow an accurate jacobian to be computed, so an ADInterfaceKernel for the dielectric boundary condition is also included. (Thanks for adding ADInterfaceKernels, Alex and Casey!)  Includes a test of the material and interface kernels with a dielectric barrier discharge under tests/surface_charge.

Before merging, I'd like to open this up to discussion regarding the jacobian testing for AD. I've included a jacobian check for the test which will fail here. I've never had success with it. In this case the test fails with quite large differences between jacobians (the test says that a "good" example is 1e-8, and I'm getting 0.999061...). The solve converges with one linear iteration and anywhere from 3 to 6 nonlinear iterations for the duration of the simulation, and the results look very close to the Comsol example I based the test on. I take that to mean AD is computing the Jacobians properly. I'm not sure what to make of this.

### CODE REVIEW

**Comment by cticenhour** [comment]
At a glance, there's a lot of good work here! That Jacobian ratio does trouble me a bit....I'd like for us to get down to at least `1e-6` here if possible. I can take a look at this input file in more depth this afternoon. The fact that your solve behavior and solution both looks good even with a "poor" Jacobian definitely is confusing. Have you run this in debug mode?

You've also beaten me to some changes I was making here and there in my AD PR (trying to get that up this week), so thanks for that! 😄

**Comment by cticenhour** [comment]
One follow-up after looking at the CIVET output - your initial Jacobian ratio looks *awesome* on the initial iteration, and then gets worse. I've had this happen in some electrothermal simulations I've been doing lately, though not quite to the extent you're seeing here. If you run the jacobian tester for multiple timesteps, do you see this same behavior (~1e-6 or less Jacobian at first and then gets to ~1 by the end of the timestep) for each one?

**Comment by keniley1** [comment]
I have not run this in debug mode.  Anything in particular I should be looking for there? I've only used that to track down seg faults, failed solves, and stuff like that, not for problems with jacobians. I'll give it a shot though. 

>  If you run the jacobian tester for multiple timesteps, do you see this same behavior (~1e-6 or less Jacobian at first and then gets to ~1 by the end of the timestep) for each one?

Nope, it looks like that only occurs on the first timestep.

**Comment by cticenhour** [comment]
> I have not run this in debug mode. Anything in particular I should be looking for there? I've only used that to track down seg faults, failed solves, and stuff like that, not for problems with jacobians. I'll give it a shot though.

I had an instance the other week where a solve looked fine in opt (if a bit slow to converge), but threw some errors (related to a mathematical operation I was doing with an ADMaterial, if memory serves) when running in debug. It was bizarre to see, and now I try to run in debug a bit more than before, to see if I get a tip-off about something I might be missing. 

> Maybe if I start with better initial conditions (e.g. something close to the periodic steady state) I'll get more appropriate behavior?

You might - I have seen an order of magnitude shift in the ratio with different initial conditions, but I am not sure it'll make up for the large difference you're seeing here. Worth a try though. 

I just noticed the initial conditions in the `ad_argon.i` input file - are these negative values due to MMS?

**Comment by keniley1** [comment]
> I just noticed the initial conditions in the ad_argon.i input file - are these negative values due to MMS?

Nope, it's just because the densities in Zapdos are log-molar densities. So if your initial condition is 10^14 m^-3, your actual IC would be ln(1e14 / 6.022e23) = -22.5 (approximately). That file doesn't actually use MMS; it's just an example of a DC discharge. 

On that note, it would be good to eventually have a distinction between examples and tests. A lot of these tests (including all of mine) are just example problems rather than "tests" to make sure the code is working properly.

**Comment by cticenhour** [comment]
> Nope, it's just because the densities in Zapdos are log-molar densities.

Oh duh - I need more coffee today 😅 

> On that note, it would be good to eventually have a distinction between examples and tests. A lot of these tests (including all of mine) are just example problems rather than "tests" to make sure the code is working properly.

I have been going back and forth on sneaking in a re-org of the tests on our next PR. I'd like to have a distinction between validation / example testing and verification (single object or single capability) tests. Working through the tests with my AD PR opened my eyes to how messy it has gotten in there. I'll go ahead with that, and we can discuss when I get that PR in.

**Comment by lindsayad** [comment]
You could have a really bad ratio if you have a function with a discontinuous derivative like an if statement or an absolute value. 

I would try running the same problem with PJFNK. If the nonlinear convergence is better with PJFNK then there is a problem with the Jacobian.

**Comment by keniley1** [comment]
> You could have a really bad ratio if you have a function with a discontinuous derivative like an if statement or an absolute value.

Unless I'm mistaken, the only places where I would have an if statement or absolute value is in the rate coefficients. I include an if statement setting any negative values to zero for the interpolated rate coefficients. I just commented that bit out and the results are identical, so I don't that's the problem. 

> I would try running the same problem with PJFNK. If the nonlinear convergence is better with PJFNK then there is a problem with the Jacobian.

What if they are almost the same? Running for a full simulation (which is about ~1300 timesteps), I'm getting pretty much identical convergence behavior for Newton and PJFNK. The average number of nonlinear steps is 4.999 for Newton and 5.077 for PJFNK. In both cases the maximum number of nonlinear iterations is 26 and the minimum is 2.

**Comment by lindsayad** [comment]
So you only see bad ratios in the initial time step? 

It’s encouraging that PJFNK and NEWTON seem to perform similarly. 

If the problem is small enough I would run with NEWTON with FDP and compare.

**Comment by cticenhour** [comment]
> If the problem is small enough I would run with NEWTON with FDP and
compare.

Would there be any scenario where NEWTON with FDP wouldn't have a "perfect"
ratio?

**Comment by cticenhour** [comment]
Beyond the discontinuity case you mentioned earlier, I mean.

**Comment by keniley1** [comment]
The problem fails with an FDP preconditioner. It makes it through 240 timesteps (corresponding to about one half cycle of the voltage pulse) before failing to converge, and the result up to that point looks similar to the SMP case. I'm really not sure why it fails. Does FDP not work well with stateful material properties or something? This problem includes surface charge as a boundary restricted stateful material property on both interfaces, which is used in the interface kernels for the potential. The interface kernels are applied on the plasma side of the interface and MatchedValueBCs are applied on the dielectric side.

**Comment by keniley1** [comment]
With the a FDP preconditioner, solve_type = NEWTON, and these petsc options: 
petsc_options_iname = '-pc_type -mat_fd_coloring_err -mat_fd_type'
petsc_options_value = 'lu 1e-8 ds'

the mean number of nonlinear steps is 4.959 (slightly lower than the NEWTON, SMP case), with max and min nonlinear steps the same as before (26 and 2, respectively).

**Comment by lindsayad** [comment]
FDP is subject to floating point roundoff error if the problem isn't well
scaled. So sometimes (actually not very uncommonly) a hand-coded Jacobian
(or one done with AD) will perform much better than a Jacobian that is
assembled using finite differences. So you certainly never want your
hand-coded (or AD) Jacobian to perform *worse* than a finite-difference
assembled one.

**Comment by lindsayad** [comment]
Let's focus on the initial timestep. Can you try playing around with values
of `-snes_test_err` and see how the ratio changes?

Try like:

-snes_test_err 1e-7
-snes_test_err 1e-6
-snes_test_err 1e-5
-snes_test_err 1e-4

**Comment by keniley1** [comment]
Alright, running with these options and varying snes_test_err: 

petsc_options_iname = '-pc_type -pc_factor_shift_type -pc_factor_shift_amount -snes_stol -snes_test_err'
petsc_options_value = 'lu NONZERO 1.e-10 0 1e-7'

Regardless of the option, the problem takes 13 nonlinear steps to solve the first timestep. At the initial nonlinear step the ratio looks like this: 

||J - Jfd||_F/||J||_F = 5.62794e-07, ||J - Jfd||_F = 8.22099e-06

and at the final one it looks like: 
||J - Jfd||_F/||J||_F = 0.998488, ||J - Jfd||_F = 1568.68

Here I'll list the values of snes_test_err, ||J - Jfd||_F/||J||_F at Nonlinear step 0, and ||J - Jfd||_F/||J||_F at Nonlinear step 13: 

**| snes_test_err |              0              |               13                |**
|       1e-7          |    5.62794e-07    |     0.998488             |
|       1e-6          |    5.63218e-06    |     0.998475             |
|       1e-5          |    0.000188384   |     0.998474             |
|       1e-4          |    0.000581815    |     0.998484            |
|       1e-3          |    0.0055933       |      0.999635            |

Also, I tried running the tester for a full simulation. It timed out after 54 timesteps, but this is what I was getting in the last timestep: 

>  0 Nonlinear |R| = ^[[32m1.153270e+00^[[39m
>   ---------- Testing Jacobian -------------
>   ||J - Jfd||_F/||J||_F = 7.87041e-07, ||J - Jfd||_F = 0.0271511
>       0 Linear |R| = ^[[32m1.153270e+00^[[39m
>       1 Linear |R| = ^[[32m9.533943e-14^[[39m
>  1 Nonlinear |R| = ^[[31m1.945777e+00^[[39m
>   ---------- Testing Jacobian -------------
>   ||J - Jfd||_F/||J||_F = 5.23996e-07, ||J - Jfd||_F = 0.0342543
>       0 Linear |R| = ^[[32m1.945777e+00^[[39m
>       1 Linear |R| = ^[[32m3.263636e-15^[[39m
>  2 Nonlinear |R| = ^[[32m3.276513e-01^[[39m
>   ---------- Testing Jacobian -------------
>   ||J - Jfd||_F/||J||_F = 6.26553e-07, ||J - Jfd||_F = 0.0347612
>       0 Linear |R| = ^[[32m3.276513e-01^[[39m
>       1 Linear |R| = ^[[32m2.720304e-16^[[39m
>  3 Nonlinear |R| = ^[[32m1.717627e-02^[[39m
>   ---------- Testing Jacobian -------------
>   ||J - Jfd||_F/||J||_F = 6.39214e-07, ||J - Jfd||_F = 0.0347683
>       0 Linear |R| = ^[[32m1.717627e-02^[[39m
>       1 Linear |R| = ^[[32m7.744010e-18^[[39m
>  4 Nonlinear |R| = ^[[32m6.439268e-05^[[39m
>   ---------- Testing Jacobian -------------
>   ||J - Jfd||_F/||J||_F = 6.39379e-07, ||J - Jfd||_F = 0.0347688
>       0 Linear |R| = ^[[32m6.439268e-05^[[39m
>       1 Linear |R| = ^[[32m2.953974e-20^[[39m
>  5 Nonlinear |R| = ^[[32m1.142299e-09^[[39m
> Nonlinear solve converged due to CONVERGED_FNORM_RELATIVE iterations 5
> Solve Converged!^[[39m

On a side note, I've been running the PetscJacobianTester through the ./run_tests option...is there a more direct way of running the jacobian tester?

**Comment by cticenhour** [comment]
You can use `-snes_test_jacobian` in the `petsc_options` parameter.

Those later timesteps are looking really good - about how far into the
simulation did the ratios start looking like this consistently?

**Comment by keniley1** [comment]
Around timestep 47 or so, although it was actually slightly increasing by the end (from 4e-7 at the minimum to 8e-7). As I said, it timed out at that point so I'm not sure if it would keep increasing. This is quite early in the simulation. The potential BC is a sinusoidal voltage, and a full period consists of something like 260 timesteps.

**Comment by keniley1** [comment]
Interestingly, if I swap out the MatchedValueBC with a PenaltyInterfaceDiffusion interface kernel, the ratio at the end of the first timestep is **much** better: 3.21051e-06.  Of course, during the second timestep it is 0.00116386...poor, but better than before. It fluctuates around during subsequent timesteps but overall seems to decrease as the simulation proceeds. 

Looks like there's some kind of interesting behavior going on with the interface. I initially thought the issue was with the surface charge so I swapped out the surface charge material and included an "ADGenericMaterialProperty" in an effort to remove surface charge from the jacobian calculation, but this had no effect on the ratio. Could there be an issue with the fact that the ADPotentialSurfaceCharge interfacekernel includes contributions on both primary and neighbor sides, and I'm also applying another interface condition on top of that?

**Comment by lindsayad** [comment]
We have a test for Jacobian accuracy of ADInterfaceKernel, so I would be
skeptical that it could be wrong...but I don't rule anything out.

Having multiple interface kernels shouldn't matter.

Wait, so the Jacobian from MatchedValueBC seems bad? That one would be a
shocker.

**Comment by keniley1** [comment]
> Wait, so the Jacobian from MatchedValueBC seems bad? That one would be a
shocker.
No, I doubt that's the case. I also tried swapping MatchedValueBC out for an AD version of and the results are the same.

Could there be something wrong with the jacobian checker? All I know is that it says my jacobian is wrong even though the solve behaves well and the results are correct. I tried a few more things, all to no effect: 

1.  Removed one of the dielectrics and added a DirichletBC, so only one interface exists. (Didn't think this would be an issue, but worth a shot.) **No effect.**
2. Removed surface charge material, interface kernel, MatchedValueBC, and second potential variable so that a single potential exists and the interface conditions are imposed naturally with no surface charge (like how it's done in mean_en.i). **No effect.**
3. Removed reaction block. **No effect.**
4. Increased initial timestep from 1e-11 to 1e-8. **Big effect!** 
**1e-8:**   ||J - Jfd||_F/||J||_F = 8.62111e-05
**1e-11:**  ||J - Jfd||_F/||J||_F = 0.998908

I don't know how to interpret this.

**Comment by lindsayad** [comment]
That last one is bizarre. What does it look like for 1e-10 and 1e-9?

The jacobian checker is very unlikely to be wrong in a fundamental way although it can definitely suffer from numerical round off error like we talked about earlier on this thread. 

Do you have any mesh adaptivity or periodic boundary conditions?

**Comment by keniley1** [comment]
> That last one is bizarre. What does it look like for 1e-10 and 1e-9?

Here's a little table of the results. The last column is the ||J - Jfd||_F/||J||_F at the final nonlinear iteration of the first timestep. 

| dt | NL Iterations  | ||J - Jfd||_F/||J||_F  |  
|---|---|---|
| 1e-12  | 12  | 0.999991  | 
| 1e-11  |  13 | 0.999061  | 
| 1e-10  | 14  | 0.760849  | 
| 1e-9  |  15 | 0.00634272  | 
| 1e-8  |  16 |  8.57935e-05 | 

> Do you have any mesh adaptivity or periodic boundary conditions?

Nope, just DirichletBCs, the Hagelaar bcs you're familiar with, and the interface kernel I've included in this PR for the potential at the interfaces. Nothing fancy going on with the mesh either. (The surface charge is a stateful material property so I don't think mesh adaptivity would work even if I wanted it to.)

**Comment by keniley1** [comment]
Anyone have any ideas about what to do with this thing? I can keep the increased timestep and just reduce the tolerance to make the test pass (1e-4 isn't horrible for something as poorly scaled as this problem...). Alternatively, I can reduce this test to a much simpler model (basically just a single charged species hitting a wall, making sure the charge buildup accumulates as expected) and relegate this more involved model to our eventual examples folder. 

I'm really not sure what's going on with the Jacobian scaling here and I don't want to ignore the problem, but I don't think the issue has much to do with the surface charge material or interface kernel.

**Comment by cticenhour** [comment]
> Anyone have any ideas about what to do with this thing? I can keep the increased timestep and just reduce the tolerance to make the test pass (1e-4 isn't horrible for something as poorly scaled as this problem...). Alternatively, I can reduce this test to a much simpler model (basically just a single charged species hitting a wall, making sure the charge buildup accumulates as expected) and relegate this more involved model to our eventual examples folder.

I would be for making a very simple model as the main test for this capability (for lightweightness, etc.) with a more involved example (that could also be a test, marked heavy). Do you have a simple model that you've been using to troubleshoot this issue that would be ok? 

> 
> I don't think the issue has much to do with the surface charge material or interface kernel.

I would tend to agree based on the troubleshooting you've performed that it's not obviously related to the physics you've added or this specific IK object. The MatchedValueBC removal *improving* things is still puzzling me. 

Regardless until our testing system at INL comes back up, we can't merge this yet since all of the webhooks that automate testing and devel -> master merges are also down.

**Comment by keniley1** [comment]
> Do you have a simple model that you've been using to troubleshoot this issue that would be ok?

No, because that would have been the smart thing to do and I avoid doing things intelligently whenever possible... I have an idea that I believe would be easy enough to make though. I'll give it a shot this week. I think I could just have a time-varying flux of particles hitting the wall to make sure the surface charge accumulates appropriately. If I make it simple enough (e.g. constant electric field, no chemistry, 1 or 2 charged species) I think I can find an analytical solution for the surface charge. 

> Regardless until our testing system at INL comes back up, we can't merge this yet since all of the webhooks that automate testing and devel -> master merges are also down.

Yeah, I figured as much. I just wanted to have this passing tests and approved so it's ready to go whenever INL gets back on its feet. I don't like having an unsolved problem hanging around when it's so maddeningly close to being done!

**Comment by cticenhour** [comment]
> No, because that would have been the smart thing to do and I avoid doing things intelligently whenever possible... I have an idea that I believe would be easy enough to make though. I'll give it a shot this week. I think I could just have a time-varying flux of particles hitting the wall to make sure the surface charge accumulates appropriately. If I make it simple enough (e.g. constant electric field, no chemistry, 1 or 2 charged species) I think I can find an analytical solution for the surface charge.

Hah! I feel that 😄 Sounds reasonable to me - I'm also crossing my fingers that the simplification will help us see whats going on here. 

> Yeah, I figured as much. I just wanted to have this passing tests and approved so it's ready to go whenever INL gets back on its feet. I don't like having an unsolved problem hanging around when it's so maddeningly close to being done!

I'm the same way. The testing issues have driven us all bonkers over here.....I have several PRs ready to go on another project (and the Zapdos AD work here) that I can't do anything with officially until this is resolved.

**Comment by keniley1** [comment]
Alright, I have officially tried everything I can think of. I couldn't really get a simpler test working. I have one with only one dielectric, but it has the same behavior as shown here.  I did streamline the surface charge material though - thanks to the power of AD, I don't need to discriminate between electrons and ions and the flux calculation is identical between them. Makes the input parameters and code much cleaner. 

I've set up the tests to pass - the exodiff should work already, but the jacobian tester I've scaled back to a tolerance 1e-4 to force it to pass. I'm out of ideas and I can't track down where the error is. What do you guys think? 

(Also, I thought updating this branch would start the test...it didn't for some reason, and my commits are logged as happening on August 3 instead of five minutes ago. Weird.)

**Comment by keniley1** [comment]
Alright, I spoke too soon. I realized there was a dumb typo in my simplified test that prevented it from working properly. I replaced the complicated DBD test with a simple one with two species ("pos" and "neg") and a varying voltage. My jacobian checker consistently shows values on the order of 1e-8, so I consider this a success! 

I think the jacobian issue might have something to do with the interpolated rate and transport coefficients for electrons, because in this example I just set constant diffusion coefficients and include no reactions. The interface kernel and material here seem to work.

I still can't get the tests restarted though. I tried "git commit --amend" and a force push, but that didn't seem to work.

**Comment by cticenhour** [comment]
I'd certainly call that a success! And if the interpolated coefficients
don't have enough derivative info, then that could be a cause of the
difference, I believe.

Regarding tests, I think that the webhooks that trigger testing need to be
modified by our system administrator - some access keys that the webhooks
use changed. I'll touch base with him on Monday and get this going.

**Comment by lindsayad** [comment]
How do you interpolate the coefficients?

**Comment by keniley1** [comment]
Same way it's always been - I use splines or linear interpolations and apply the chain rule for the AD derivative terms. It could definitely be a source of error if I am doing that incorrectly. The diffusion coefficient looks like this (see materials/ADGasElectronMoments): 

`_diffem[_qp].value() =
      _diff_interpolation->sample(std::exp(_mean_en[_qp].value() - _em[_qp].value())) *
      _time_units;

  _diffem[_qp].derivatives() =
      _diff_interpolation->sampleDerivative(std::exp(_mean_en[_qp].value() - _em[_qp].value())) *
      std::exp(_mean_en[_qp].value() - _em[_qp].value()) *
      (_mean_en[_qp].derivatives() - _em[_qp].derivatives()) * _time_units;`

**Comment by lindsayad** [comment]
That looks like a good chain rule to me. I dunno man. At least you've got
this simple test working. I say bravo to that

**Comment by cticenhour** [comment]
Alright @keniley1 I've reset the webhook and restarted testing. Would you say this is ready for review after these pass, or was there anything else you'd like to add?

**Comment by cticenhour** [comment]
Looks like you've got some trailing whitespace in two inputs: 

```
ERROR: The following files contain trailing whitespace after applying your patch:
	tests/surface_charge/dbd_test.i
	tests/surface_charge/interface_test.i

Run the "delete_trailing_whitespace.sh" script in your $MOOSE_DIR/scripts directory.
```

**Comment by keniley1** [comment]
I always have trailing whitespaces and clang formatting fails on my first try...

Yeah, In terms of functionality I think this is good as it is. The only obvious thing we might want to change is renaming the material and interface kernel to remove the "AD" since we're planning on making everything AD by default, but that's not a huge deal.

**Comment by cticenhour** [comment]
> I always have trailing whitespaces and clang formatting fails on my first try...

Are you using Atom? If so, I highly recommend going into the _whitespace_ core package settings (Settings --> packages --> search for _whitespace_) and deactivate "Ignore Whitespace on Current Line". I no longer get whitespace testing errors. 

> The only obvious thing we might want to change is renaming the material and interface kernel to remove the "AD" since we're planning on making everything AD by default, but that's not a huge deal.

I can update this when I finally clean up my Zapdos AD PR for merging this week, so I don't mind leaving it as is for now.

**Comment by keniley1** [comment]
Nope, I use Vim. I have clang formatting on vim that I use and it works really well, but I always seem to forget to use it on at least one file whenever I submit a PR. I'm sure there's a whitespace correction function in Vim as well so I'll search around. 

> I can update this when I finally clean up my Zapdos AD PR for merging this week, so I don't mind leaving it as is for now.
Alright, sounds good. Let me know if you notice anything else that needs fixing though!

**Comment by cticenhour** [comment]
Ah ok - I imagine there is. 

Since we've been sitting on this a while, I'll look over this as soon as I can - attending GEC sessions and a few INL meetings all day, so it may not be until later this afternoon, but I promise I'll get to it today 😄

File: include/interfacekernels/ADPotentialSurfaceCharge.h
Comment (cticenhour) [inline include/interfacekernels/ADPotentialSurfaceCharge.h:16]: Could you make this class description comment a bit more specific to this class? This looks like the same one from the `ADCoupledInterfacialSource` classes in MOOSE.

File: tests/automatic_differentiation/ad_argon.i
Comment (cticenhour) [inline tests/automatic_differentiation/ad_argon.i:5]: Several instances of lingering commented lines from troubleshooting in this file....could this be cleaned up since the test has been re-golded? Or if it needs to remain to illustrate something, include a descriptive comment.

File: tests/automatic_differentiation/ad_argon.i
Comment (cticenhour) [inline tests/automatic_differentiation/ad_argon.i:771]: Just a comment - I really appreciate the extended description comments / caveats here. Will be a huge help for new users!

File: include/actions/AddDriftDiffusionAction.h
Comment (cticenhour) [inline include/actions/AddDriftDiffusionAction.h:59]: Could you remove this old comment while making these other changes?

File: src/actions/AddDriftDiffusionAction.C
Comment (cticenhour) [inline src/actions/AddDriftDiffusionAction.C:723]: Clean-up stray comments if no longer needed.

**Review by cticenhour — CHANGES_REQUESTED** [review]
Good work - most of my comments are readability and quality of life improvements.

File: tests/automatic_differentiation/ad_argon.i
Comment (cticenhour) [reply tests/automatic_differentiation/ad_argon.i:5]: I just commented on the first instance, but this happens several times in this file.

File: include/actions/AddDriftDiffusionAction.h
Comment (keniley1) [reply include/actions/AddDriftDiffusionAction.h:59]: Got it! 👍

File: tests/automatic_differentiation/ad_argon.i
Comment (keniley1) [reply tests/automatic_differentiation/ad_argon.i:5]: Yeah, sorry - I leave a mess sometimes. I'll clean this up.

**Comment by keniley1** [comment]
I believe I addressed all of the comments! Thanks for looking through this.

**Comment by cticenhour** [comment]
Those darn whitespaces! 😄 

```
ERROR: The following files contain trailing whitespace after applying your patch:
	include/interfacekernels/ADPotentialSurfaceCharge.h

Run the "delete_trailing_whitespace.sh" script in your $MOOSE_DIR/scripts directory.
```

**Review by cticenhour — APPROVED** [review]
Thanks for implementing this! I'll merge when the testing completes.

**Comment by keniley1** [comment]
No problem! It's definitely a useful thing to have - dielectrics are really common in plasmas. 

One day I will submit a PR and it will have no trailing whitespaces, no clang formatting issues, and every test will past on the first try. One day...

**Pull request merged**
